src/dune/build_system.ml

```diff
@@ -1517,6 +1517,27 @@ end = struct
               |> Result.map_error ~f:report |> ignore
             | _ -> ()
           in
+          let () =
+            (* Remove write permissions on targets. A first theoretical reason
+               is that the build process should be a computational graph and
+               targets should not change state once built. A very practical
+               reason is that enabling the cache will remove write permission
+               because of hardlink sharing anyway, so always removing them
+               enables to catch mistakes earlier. *)
+            let f (path, _) =
+              (* FIXME: searching the dune version for each single target seems
+                 way suboptimal. This information could probably be stored in
+                 rules directly. *)
+              let _, src_dir = Path.Build.extract_build_context_dir_exn path in
+              let dir = File_tree.nearest_dir src_dir in
+              let version =
+                File_tree.Dir.project dir |> Dune_project.dune_version
+              in
+              if version >= (2, 4) then
+                Path.Build.chmod ~mode:0x222 ~op:`Remove path
+            in
+            List.iter ~f targets
+          in
```

Comment:
We already stat all the targets in `compute_targets_digest_or_raise_error`, via `Cached_digest.refresh`. It doesn't feel optimal to start the targets twice systematically all the time. What about renaming `refresh` to `refresh_and_chmod` and doing both at once?